File: Backend/src/main/java/com/syntaxcore/workforce/service/ProjectService.java
package com.syntaxcore.workforce.service;

import com.syntaxcore.workforce.dto.request.ProjectRequest;
import com.syntaxcore.workforce.dto.response.PagedResponse;
import com.syntaxcore.workforce.dto.response.ProjectResponse;
import com.syntaxcore.workforce.enums.ProjectStatus;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.UUID;

public interface ProjectService {
    ProjectResponse createProject(ProjectRequest request, String creatorEmail);
    ProjectResponse updateProject(UUID id, ProjectRequest request);
    ProjectResponse getProject(UUID id);
    PagedResponse<ProjectResponse> getAllProjects(ProjectStatus status, UUID managerId, String search, Pageable pageable);
    void deleteProject(UUID id);
    List<ProjectResponse> getMyProjects(String userEmail);
    ProjectResponse addMember(UUID projectId, UUID userId);
    ProjectResponse removeMember(UUID projectId, UUID userId);
}

File: Frontend/src/app/core/services/project.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Project, PagedResponse, ProjectStatus } from '../models';
import { environment } from '../../../environments/environment';

@Injectable({ providedIn: 'root' })
export class ProjectService {
  private readonly API = `${environment.apiUrl}/projects`;
  constructor(private http: HttpClient) {}

  createProject(request: Partial<Project>): Observable<Project> {
    return this.http.post<Project>(this.API, request);
  }

  getAllProjects(filters: { status?: ProjectStatus; search?: string; page?: number; size?: number } = {}): Observable<PagedResponse<Project>> {
    let params = new HttpParams();
    Object.entries(filters).forEach(([k, v]) => { if (v !== undefined) params = params.set(k, String(v)); });
    return this.http.get<PagedResponse<Project>>(this.API, { params });
  }

  getProject(id: string): Observable<Project> {
    return this.http.get<Project>(`${this.API}/${id}`);
  }

  updateProject(id: string, request: Partial<Project>): Observable<Project> {
    return this.http.put<Project>(`${this.API}/${id}`, request);
  }

  deleteProject(id: string): Observable<void> {
    return this.http.delete<void>(`${this.API}/${id}`);
  }

  getMyProjects(): Observable<Project[]> {
    return this.http.get<Project[]>(`${this.API}/my-projects`);
  }

  addMember(projectId: string, userId: string): Observable<Project> {
    return this.http.post<Project>(`${this.API}/${projectId}/members/${userId}`, {});
  }

  removeMember(projectId: string, userId: string): Observable<Project> {
    return this.http.delete<Project>(`${this.API}/${projectId}/members/${userId}`);
  }
}
